File: Linked_List_free_code_camp/5.linked_list_get_value.js
class Node {
    constructor(val, next = null) {
        this.val = val;
        this.next = next;
    }
}

const head =
    new Node(1,
        new Node(2,
            new Node(3,
                new Node(4,
                    new Node(5, null)))));

// A -> B -> C -> D -> null

const getNodeValue = (head, index) => {
    let ans = null;
    let i = 0;

    while (head && i <= index) {
        if (i == index) {
            ans = head.val;
            break;
        }
        head = head.next;
        i++;
    }

    return ans;
};

console.log(getNodeValue(head, 2));


const getNodeValueRecursive = (head, index) => {
    if (head === null) return null;
    console.log(head.val);
    if (index === 0) return head.val;
    let ans = null;

    ans = getNodeValueRecursive(head.next, index - 1);

    return ans;
};

console.log(getNodeValueRecursive(head, 1));

File: Linked_List_free_code_camp/5.linked_list_get_value.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    Node* next;

    Node(int val, Node* next = NULL) {
        this->val = val;
        this->next = next;
    }
};

void print_v(vector<int> values) {
    for (auto e : values) {
        cout << e << " ";
    }
    cout << endl;
}

int getNodeValue(Node* head, int index) {
    Node* t = head;

    while (t && index--) {
        t = t->next;
    }

    return t->val;
};

int getNodeValueRecursive(Node* head, int index) {
    if (head == NULL) return -1;
    if (index == 0) return head->val;

    int faith = getNodeValueRecursive(head->next, index - 1);

    return faith;
};


int main() {

    Node* head =
        new Node(10,
            new Node(20,
                new Node(30,
                    new Node(40))));

    cout << getNodeValue(head, 2) << endl;
    cout << getNodeValueRecursive(head, 2) << endl;

    return 0;
}